Use list comprehesions to do one thing. Create lists.

In [21]:
symbols = "$¢£¥€¤"
codes = [ord(symbol) for symbol in symbols]
print(codes)

[36, 162, 163, 165, 8364, 164]


A generator expression is enclosed in parentheses and saves memory by only providing one item to a function call at a time. We can use them to do the same thing. I'm not quite sure what they do but that is covered in more detail in Chapter 4.

In [8]:
[x for x in (ord(symbol) for symbol in symbols)]

[36, 162, 163, 165, 8364, 164]

 Tuples can be used as records as their position gives them meaning.

We can unpack tuples too!

In [12]:
compass_points = ("North", "South", "East", "West")
North, South, East, West = compass_points
North

'North'

Which is often useful for swapping variables. The * prefix unpacks variables.

In [14]:
t = (20, 8)
quotient, remainder = divmod(*t)

We can also use only part of this output by unpacking to a dummy variable `_`.

In [18]:
_, _ = divmod(*t)

The `*` operator can also be used to grab excess items from a function call which also works for parallel assignment.

In [20]:
a, b, *rest = range(5)

[2, 3, 4]

In [24]:
*rest, a, b = range(5)

Tuple unpacking also works on nested structures if we match the nesting structures in the assigment.

In [27]:
city, country_code, population_in_millions, (latitude, longitude) = (
    "New York-Newark",
    "US",
    20.104,
    (40.808611, -74.020386),
)

Named tuples allow the creation of simple classes.

In [8]:
from collections import namedtuple

City = namedtuple("City", "name country population coordinates")
tokyo = City("Tokyo", "JP", 36.933, (35.689722, 139.691667))
tokyo.name

'Tokyo'

# Slices


In [11]:
l = list(range(10))

Interesting note that Python slice convention:
* Is related to the length of the list returned rather than the index.


In [13]:
l[:3]

[0, 1, 2]

* Allows easy computation of the list length.


In [14]:
len(l[1:4]) == 3

True

* Enables splitting of lists without over lap.

In [15]:
l[3:]

[3, 4, 5, 6, 7, 8, 9]

Slice items can be used rather than the colon operator.

In [16]:
my_slice = slice(1, 2)
l[my_slice]

[1]

We can assign to slices too

In [19]:
l[1:2] = [-1]
l

[0, -1, 2, 3, 4, 5, 6, 7, 8, 9]

A common pitfall is that...

In [20]:
[["_"] * 3] * 3

[['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]

is equivalent to...

In [23]:
row = ['_'] * 3
weird_board = [row for _ in range(3)]
weird_board[0][0] = 1
weird_board

[[1, '_', '_'], [1, '_', '_'], [1, '_', '_']]

and copies the same row three times.

A strange example is given

In [25]:
t = (1, 2, [30, 40])
t[2] += [50, 60]

TypeError: 'tuple' object does not support item assignment

In [26]:
t

(1, 2, [30, 40, 50, 60])

By convention, in place changes are signalled by functions and methods that return nothing.

`sorted` and `sort` create a copy and in-lace sorting respectively.

Arrays should be used to replace lists if all entries are the same and lists are supported.

In [27]:
from array import array
from random import random
floats = array('d', (random() for i in range(10**7)))

In [28]:
floats[-1]

0.4427969295213573

Memory views can be used to change elements of large data sets.

In [2]:
from array import array
numbers = array('h',[-2,-1,0,1,2])
memv = memoryview(numbers)
len(memv)
memv[0]


-2

The code is quite understandable upto here.

In [5]:
memv_oct = memv.cast('B')
memv_oct.tolist()

[254, 255, 255, 255, 0, 0, 1, 0, 2, 0]

In these last few lines...

The `cast` method of a memoryview object returns a new memoryview object that points to the same memory block, but with a different format. In this case, the format is changed from signed short integers ('h') to unsigned char ('B'). This changes how the data in the memory block is interpreted.

When the memoryview object is converted to a list using the tolist method, each element of the original array is represented as two elements in the new list. This is because a signed short integer takes up two bytes of memory, while an unsigned char takes up only one byte. The values in the new list are obtained by splitting each signed short integer into two bytes and interpreting each byte as an unsigned char.

For example, the first element of the original array is -2. In binary, this is represented as 1111111111111110. When this value is split into two bytes, we get 11111111 and 11111110, which are 255 and 254 in decimal, respectively. These values are then interpreted as unsigned chars and added to the new list.

Is there anything else you would like to know?

In [6]:
print(bytes(memv_oct[1]))

b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x

In [7]:
numbers

array('h', [-2, -1, 0, 1, 2])